sw/device/lib/dif/dif_i2c.h

```diff
@@ -0,0 +1,483 @@
+
+/**
+ * Pushes a raw write entry onto the FMT FIFO, consisting of a byte and format
+ * flags. This function can be called in sequence to enqueue an I2C
+ * transmission.
+ *
+ * Callers should prefer `dif_i2c_write_byte()` instead, since that function
+ * provides clearer semantics. This function should only really be used for
+ * testing or troubleshooting a device.
+ *
+ * @param i2c An I2C device.
+ * @param byte The value to push onto the FIFO.
+ * @param flags The flags to use for this write.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_write_byte_raw(const dif_i2c_t *i2c, uint8_t byte,
+                                        dif_i2c_fmt_flags_t flags);
+/**
+ * Available formatting codes for `dif_i2c_write_byte_raw()`.
+ *
+ * Each code describes how to interpret the `byte` parameter, referred to below
+ * as "the byte".
+ *
+ * It is the caller's responsibility to observe the state transitions in the
+ * comments below.
+ */
+typedef enum dif_i2c_fmt {
+  /**
+   * Start a transaction. This sends a START signal followed by the byte.
+   * The byte sent will form (potentially part of) the target address for the
+   * transaction.
+   *
+   * May be followed by any format code.
+   */
+  kDifI2cFmtStart,
+  /**
+   * Transmit byte. This simply sends the byte. It may need to be used in
+   * conjunction with `Start` to send a multi-byte target address.
+   *
+   * May be followed by any format code.
+   */
+  kDifI2cFmtTx,
+  /**
+   * Transmit byte and stop. This sends the byte, and then sends a stop
+   * signal, completing a transaction.
+   *
+   * Only `Start` may follow this code.
+   */
+  kDifI2cFmtTxStop,
+  /**
+   * Request `n` bytes, where `n` is the byte interpreted as an unsigned
+   * integer; a byte value of `0` will be interpreted as requesting `256`
+   * bytes. This will NAK the last byte.
+   *
+   * Only `Start` may follow this code (this code does not stop a transaction;
+   * see `RxStop`).
+   */
+  kDifI2cFmtRx,
+  /**
+   * Request `n` bytes, same as `Rx`, but ACK the last byte so that more data
+   * can be requested.
+   *
+   * May be followed by `RxContinue`, `Rx`, or `RxStop`.
+   */
+  kDifI2cFmtRxContinue,
+  /**
+   * Request `n` bytes, same as `Rx`, but, after NAKing the last byte, send a
+   * stop signal to end the transaction.
+   *
+   * Only `Start` may follow this code.
+   */
+  kDifI2cFmtRxStop,
+} dif_i2c_fmt_t;
+
+/**
+ * Pushes a write entry onto the FMT FIFO, consisting of a byte and a format
+ * code. This function can be called in sequence to enqueue an I2C
+ * transmission.
+ *
+ * @param i2c An I2C device.
+ * @param byte The value to push onto the FIFO.
+ * @param code The code to use for this write.
+ * @param supress_nak_irq Whether to supress the NAK IRQ for this one byte.
+ *        May not be used in combination with `Rx` codes.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_write_byte(const dif_i2c_t *i2c, uint8_t byte,
```

Comment:
I don't see tests for this. Please file an issue to add them after the fact.

Reply:
Done.